### DISCUSSION TITLE
Embedding BIRT Reporting Engine in JavaFX stand alone application

### DISCUSSION THREAD

**Discussion by markokue** [eclipse-birt/birt#1354] [Q&A]
I am developing a basic stand alone JavaFX Application with an embedded database and Hibernate. My goal is to embed BIRT Reporting Engine to dynamically generate reports. It is a Maven project.

Unfortunately the instructions given on BIRT website are not clear (at least to me) and seem to be outdated.

How can I embed the Reporting Engine in my project?

Any lead is highly appreciated.
Thanks you.

**Comment by wimjongman** [comment]
What do you want to embed, the running of reports or also the design of reports? Please note that the report designer is based on SWT and not on JavaFX.

Is your application an OSGi-based application?

**Reply by markokue** [reply]
Dear wimjongman. Thanks for your reply. I achieved the integration of the Reporting Engine itself in my project. However I don't understand why a local reference is still required for the Report Engine. (Btw. I had additionally include one more dependency (itext) to be able to save my pdf report. That library is not included in the BIRT Reporting Engine download package.)

Here is my code I am using:

```
config = new EngineConfig();
            config.setBIRTHome("/Users/xxx/Downloads/birt-runtime-4.14.0-202306241523/ReportEngine");
            config.setLogConfig("/Users/xxx/Downloads/", Level.FINEST);
            Platform.startup(config);
            final IReportEngineFactory FACTORY = (IReportEngineFactory) Platform
                    .createFactoryObject(IReportEngineFactory.EXTENSION_REPORT_ENGINE_FACTORY);
            engine = FACTORY.createReportEngine(config);
            engine.changeLogLevel(Level.WARNING);
```

            IReportRunnable report = engine.openReportDesign("/Users/xxx/Downloads/new_report.rptdesign");
            IRunAndRenderTask task = engine.createRunAndRenderTask(report);
            //IRenderOption options = new RenderOption();
            task.getAppContext().put("student", student);
            PDFRenderOption PDF_OPTIONS = new PDFRenderOption();
            PDF_OPTIONS.setOutputFileName("/Users/xxx/Downloads/printout.pdf");
            PDF_OPTIONS.setOutputFormat("pdf");
            task.setRenderOption(PDF_OPTIONS);
            task.run();
            task.close();`

My application generates dynamically reports based on a rptdesign file. It is a standalone JFX based application.
Thanks.

**Reply by wimjongman** [reply]
Ok, so what is your remaining problem?

**Reply by markokue** [reply]
Why is that still required `config.setBIRTHome("/Users/xxx/Downloads/birt-runtime-4.14.0-202306241523/ReportEngine");` since all the the required libraries have been included in the project?

**Reply by 4r73m190r0s** [reply]
Did you find the answer to your question? It's also strange to me that we need to set `BIRT_HOME` even though we imported Report Engine libraries into the project.

**Reply by TheLaserNinja** [reply]
You have to remember, through most of BIRT's life, it was developed by a commercial entity that built this for their product. It wasn't until BIRT took off with the old portal and books we put out that people didn't want to use OSGi. 
It was a workaround from before we had a standalone report engine. Since the iServer and iHub used OSGi for initially, we only had the OSGi implementation for several years. The method we have now was a workaround, and IIRC, was done for both backwards compatibility and as some weird workaround for the classloader to find the necessary classes to work without OSGi. I'm not sure that it is still required.

**Comment by hvbtup** [comment]
I don't think this is required if your program doesn't use OSGI.

**Reply by hvbtup** [reply]
Did you read the answer I gave?
